### DISCUSSION TITLE
SNOW AND SNOW2 NOT GIVING THE SAME VALUE OF PORES, THROATS AND PERMEABILITY

### DISCUSSION THREAD

**Discussion by JahanzebTariq** [PMEAL/OpenPNM#2637] [Q&A]
DEAR PROFESSOR, I have been able to run the code and extract network through open PNM and porespy , but **the issue I face is that I am not getting the same number of throats, pores and most importantly the permeability value**. I humbly request you to please solve my issue. 
Thank you

**Comment by jgostick** [comment]
Yes, that's why we called it snow2, since there are some differences in the code. How different are your permeability values?

**Reply by JahanzebTariq** [reply]
X: [999.4799] mD
Y: [976.0905] mD
Z: [960.0699] mD

These are the original once

X: [14.2838] mD

This is what I'm getting..

**Reply by JahanzebTariq** [reply]
This is the original code:
import numpy as np
import openpnm as op
import porespy as ps
import matplotlib.pyplot as plt
np.set_printoptions(precision=4)
np.random.seed(10)
%matplotlib inline

resolution = 2.25e-6
name = 'Bentheimer'

# Read input RAW file
raw_file = np.fromfile('Sandstones/'+name+'_2d25um_binary.raw', dtype=np.uint8)
im = (raw_file.reshape(1000,1000,1000))
im = im==0;

**CONFIRM IMAGE AND CHECK IMAGE POROSITY**
#NBVAL_IGNORE_OUTPUT
fig, ax = plt.subplots(1, 3, figsize=(12,5))
ax[0].imshow(im[:, :, 100]);
ax[1].imshow(ps.visualization.show_3D(im[:250,:250,:250]));
ax[2].imshow(ps.visualization.sem(im[:250,:250,:250]));
ax[0].set_title("Slice No. 100 View");
ax[1].set_title("3D Sketch");
ax[2].set_title("SEM View");

print(ps.metrics.porosity(im))
print(im.shape)
print(im.dtype)

**Extract pore network using SNOW algorithm in PoreSpy**
net = ps.networks.snow(im=im, voxel_size=resolution)

**IMPORT NETWORK IN PNM**
ws = op.Workspace()
proj = op.Project()
pn = op.network.GenericNetwork(name=name, project=proj)
pn.update(net)  # Fills 'pn' with data from 'net'
print(pn)

CHECK NETWORK HEALTH
h = pn.check_network_health()
op.topotools.trim(network=pn, pores=h['trim_pores'])
h = pn.check_network_health()
print(h)

**ASSIGNING GEOMETRY**
geo = op.geometry.GenericGeometry(network=pn, pores=pn.Ps, throats=pn.Ts)

**ASSIGNING PHASE**
water = op.phases.Water(network=pn)

**ASSIGNING PHYSCIS**
phys_water=op.physics.GenericPhysics(network=pn, phase=water, geometry=geo)
R = geo['throat.diameter']/2.0;
L = geo['throat.length'];
phys_water['throat.hydraulic_conductance'] = (np.pi*(R**4))/(8*water['pore.viscosity'].max()*L)


**CALCULATE EFFECTIVE PERMEABILITY**
A = (1000*1000) *resolution**2
L = 1000 * resolution
mu = water['pore.viscosity'].max()
Pressure = 10e3;
delta_P = Pressure - 0
#X
perm = op.algorithms.StokesFlow(network=pn, project=proj)
perm.setup(phase=water)
perm.set_value_BC(pores=pn.pores('top'), values=0)
perm.set_value_BC(pores=pn.pores('bottom'), values=Pressure)
perm.run()
Q = perm.rate(pores=pn.pores('bottom'), mode='group')
K = Q * L * mu / (A * delta_P)
print('X:', K/0.98e-12*1000, 'mD')

#y
perm = op.algorithms.StokesFlow(network=pn, project=proj)
perm.setup(phase=water)
perm.set_value_BC(pores=pn.pores('front'), values=0)
perm.set_value_BC(pores=pn.pores('back'), values=Pressure)
perm.run()
Q = perm.rate(pores=pn.pores('back'), mode='group')
K = Q * L * mu / (A * delta_P)
print('Y:', K/0.98e-12*1000, 'mD')

#z
perm = op.algorithms.StokesFlow(network=pn, project=proj)
perm.setup(phase=water)
perm.set_value_BC(pores=pn.pores('left'), values=0)
perm.set_value_BC(pores=pn.pores('right'), values=Pressure)
perm.run()
Q = perm.rate(pores=pn.pores('right'), mode='group')
K = Q * L * mu / (A * delta_P)
print('Z:', K/0.98e-12*1000, 'mD')

**Reply by JahanzebTariq** [reply]
**THIS IS THE UPDATED CODE BY ME:**

``` python
import numpy as np
import openpnm as op
import porespy as ps
import matplotlib.pyplot as plt
np.set_printoptions(precision=4)
np.random.seed(10)
%matplotlib inline

resolution = 2.25e-6
name = 'Bentheimer'

raw_file = np.fromfile('Sandstones/'+name+'_2d25um_binary.raw', dtype=np.uint8)
im = (raw_file.reshape(1000,1000,1000))
im = im==0;
``` 


CONFIRM IMAGE AND CHECK IMAGE POROSITY

``` python
#NBVAL_IGNORE_OUTPUT
fig, ax = plt.subplots(1, 3, figsize=(12,5))
ax[0].imshow(im[:, :, 100]);
ax[1].imshow(ps.visualization.show_3D(im[:250,:250,:250]));
ax[2].imshow(ps.visualization.sem(im[:250,:250,:250]));
ax[0].set_title("Slice No. 100 View");
ax[1].set_title("3D Sketch");
ax[2].set_title("SEM View");

print(ps.metrics.porosity(im))
print(im.shape)
print(im.dtype)

snow = ps.networks.snow2(im, voxel_size=resolution)

pn = op.io.network_from_porespy(snow.network)

h = op.utils.check_network_health(pn)
op.topotools.trim(network=pn, pores=h['disconnected_pores'])
h = op.utils.check_network_health(pn)
print(h)

pn['pore.diameter'] = pn['pore.equivalent_diameter']
pn['throat.diameter'] = pn['throat.inscribed_diameter']
pn['throat.spacing'] = pn['throat.total_length']

from pprint import pprint
mods = op.models.collections.geometry.spheres_and_cylinders
pprint(mods)

pn.add_model_collection(mods)
pn.regenerate_models()
print(pn)

water = op.phase.Water(network=pn)
print(water)

water.add_model_collection(op.models.collections.physics.standard)
R = pn['throat.diameter']/2.0;
L = pn['throat.length'];
water['throat.hydraulic_conductance'] = (np.pi*(R**4))/(8*water['pore.viscosity'].max()*L)
water.regenerate_models()

``` 

## Perform Stokes flow to find Permeability coefficient

``` python
A = (1000*1000) *resolution**2
L = 1000 * resolution
mu = water['pore.viscosity'].max()
Pressure = 10e3;
delta_P = Pressure - 0

sf = op.algorithms.StokesFlow(network=pn, phase=water)
sf.set_value_BC(pores=pn.pores('xmin'), values=0)
sf.set_value_BC(pores=pn.pores('xmax'), values=Pressure)
sf.run()

Q = sf.rate(pores=pn.pores('xmax'), mode='group')
K = Q * L * mu / (A * delta_P)
print('X:', K/0.98e-12*1000, 'mD')

```

**Reply by jgostick** [reply]
This is a lot of code to sort through.  Before I do that, a few questions:
1. What is the actual, experimental value for this material or image?
2. How did you treat the boundary pores on the snow original?

**Reply by JahanzebTariq** [reply]
The experimental value of the **porosity is 22.64% and permeability is 386 mD**

**Reply by JahanzebTariq** [reply]
sir I used the same code as it is I did not do any thing to it.. please guide me through this

**Reply by JahanzebTariq** [reply]
I should also mention that I am replicating the original code and using the same conditions given in the original code.. so please do tell me where am I going wrong

**Reply by jgostick** [reply]
So snow2 is high by a factor of 3. This is not ideal, but is not too far off actually.  There are probably a few things you can tweak to get it more in line.  I'm sorry but I don't have the time to comb through your code right now.  Have you looked at the code in the link i sent?  Try running that code on your image directly?

**Reply by JahanzebTariq** [reply]
sure professor.. but that is for Berea sandstone.. I wanted to run it on Bentheimer sandstone.. and please could you please tell me where did the script writer got the formation factor and the permeability value in the x,y,z directions. i know this is a silly question but i would be glad if you could please guide me..

**Reply by JahanzebTariq** [reply]
@jgostick, Professor I did as you directed me to do.. It resulted in a permeability value of **1975.52 mD** when I replaced **pyramids and cones with cones and cylinders** which is much better than the previous value I got, but still this is not the value given in the paper. The original published code is giving a value of **999.479 mD**. **Also I am not being able to use spheres and cylinders which I think would give more better result as it reuslts in an error that throat diameter is greater than pore diameter**
Please I humbly request you to guide me through this. Thanks is advance

```python
import openpnm as op
import porespy as ps
import numpy as np
import os
from pathlib import Path

# Read input RAW file
resolution = 2.25e-6
name = 'Bentheimer'
raw_file = np.fromfile('Downloads/'+name+'_2d25um_binary.raw', dtype=np.uint8)
im = (raw_file.reshape(1000,1000,1000))
im = im==0;

# %% Perform extraction
snow = ps.networks.snow2(im, voxel_size=resolution, boundary_width=[3,0,0], accuracy='standard')

# %% Open network in OpenPNM
pn = op.io.network_from_porespy(snow.network)

del pn.models['pore.diameter']
del pn.models['throat.diameter']
del pn.models['throat.spacing']
del pn.models['throat.coords']

pn['pore.diameter'] = pn['pore.equivalent_diameter']
pn['throat.diameter'] = pn['throat.inscribed_diameter']
pn['throat.spacing'] = pn['throat.total_length']
pn['throat.coords'] = pn['throat.global_peak']

mod = op.models.geometry.hydraulic_size_factors.cones_and_cylinders
pn.add_model(propname='throat.hydraulic_size_factors',model=mod)
mod = op.models.geometry.diffusive_size_factors.pyramids_and_cuboids
pn.add_model(propname='throat.diffusive_size_factors',model=mod)

pn.regenerate_models()
h = op.utils.check_network_health(pn)
op.topotools.trim(network=pn, pores=h['disconnected_pores'])

water = op.phase.Water(network=pn)

water.add_model_collection(op.models.collections.physics.standard)
R = pn['throat.diameter']/2.0;
L = pn['throat.spacing'];
water['throat.hydraulic_conductance'] = (np.pi*(R**4))/(8*water['pore.viscosity'].max()*L)
water.regenerate_models()

water['pore.viscosity'].max()

# %% Perform Stokes flow to find Permeability coefficient
Pressure = 10e3
dP = Pressure-0
L = 1000*resolution
A = (1000000)*resolution**2
mu = water['pore.viscosity'].max()
inlet=pn.pores('xmin')
outlet=pn.pores('xmax')

sf = op.algorithms.StokesFlow(network=pn, phase=water)
sf.set_value_BC(pores=inlet, values=Pressure)
sf.set_value_BC(pores=outlet, values=0)
sf.run()

Q = sf.rate(pores=pn.pores('xmin'),mode ='group')
K= Q*L*mu / (A*dP)
print('x', K/0.98e-12*1000,'mD')
```

**Reply by ma-sadeghi** [reply]
@JahanzebTariq I'll let @jgostick follow up, but a general comment:

The permeability you're getting is off by a factor of 2, which is probably as good as it gets from a pore network model (PNM). With PNM, you shouldn't expect to get very accurate results comparable to those of a direct numerical simulation on voxel images. It'd be amazing if that were the case, but not there yet. We've published a [paper](https://doi.org/10.1016/j.cageo.2022.105086) on estimating the diffusive size factor of conduits in extracted PNMs using machine learning, and it was shown to give relatively accurate predictions, much better than those you get using the typical size factor models in the literature (including those implemented in OpenPNM). However, it's only for "diffusive size factor". [Here](https://porespy.org/examples/networks/tutorials/using_diffusive_size_factor_AI_with_snow.html)'s an example on how to use it.

**Reply by JahanzebTariq** [reply]
@ma-sadeghi thank for your reply. Sir the value **999.479mD was calculated using PNM** by the author using the same dataset but **using SNOW algorithm**.. **I used SNOW2**.. please can you point out more of the steps I could take to make this value come more close to the original in the paper..

**Reply by JahanzebTariq** [reply]
and I should also mention that the **experimental value of the porosity is 22.64% and permeability is 386 mD**.. The published results were **already up by a factor of 3, making my obtained results 5 times greater than the original one**...  please help me sort this out at least that I get the published results. 
Thank you

**Reply by ma-sadeghi** [reply]
@JahanzebTariq I looked at your code and something doesn't seem right: why are you hard-coding the hydraulic conductance? OpenPNM should already calcualte it for you.

```python
water['throat.hydraulic_conductance'] = (np.pi*(R**4))/(8*water['pore.viscosity'].max()*L)
```

**Reply by JahanzebTariq** [reply]
@ma-sadeghi ..sorry for replying so late.. Sir as I mentioned, **I am replicating a paper so tried to do each and every step coded. as it was coded in the previous version so i thought so hard code it**.. I'll share with you the results after cutting this step hope this time I get it right

**Reply by ma-sadeghi** [reply]
@JahanzebTariq I understand, but I think your main goal should be to match the experimental value. SNOW algorithm has evolved since then, so at least I don't see any practical point in trying to exactly replicate the paper.

**Reply by JahanzebTariq** [reply]
@ma-sadeghi .. Sir after doing as you directed I got a value of **[1760.018] mD** in the x direction

**Reply by JahanzebTariq** [reply]
@ma-sadeghi .. and sir 1 more thing I wanted to point out is SNOW2 algorithm is giving me **more numbers of pores and throats than SNOW**.
Further I wanted to ask that I am not being able to use spheres and cylinders as the message pops of **throat diameter being greater than the pore diameter**... any possible solution on this..?

**Reply by ma-sadeghi** [reply]
@JahanzebTariq Please do one more modification: try "pore.inscribed_diameter" instead of "pore.equivalent_diameter".

**Reply by JahanzebTariq** [reply]
@ma-sadeghi ... OMG sir.. You are so great.. I got **721.3061 mD** after trying "pore.inscribed_diameter" instead of "pore.equivalent_diameter".. This is more closer to the experimental result.

**Reply by JahanzebTariq** [reply]
@ma-sadeghi .. Excuse me sir.. could you kindly tell me the reason behind switching the diameters..
Thank you

**Reply by jgostick** [reply]
It's in the name:  Inscribed diameter is the diameter of a sphere fits entirely withing the pore region, whereas equivalent diameter is the diameter of a spheres with the same volume as the region.  Inscribed will always be smaller.

**Comment by jgostick** [comment]
Also, I will point you to [this script](https://github.com/PMEAL/OpenPNM/blob/dev/tests/integration/PoreSpyIO_on_Berea.py), which is part of our test suite, that ensures the extracted network gives correct Permeability and Formation Factor values for Berea:

**Reply by JahanzebTariq** [reply]
Dear Professor I also took help from this script.. I am a MS student and currently in my research phase so I am trying to replicate a paper whose code was the first one. The second is my code.. please could you check from the portion from where I applied the snow2 algorithm.

**from previous paper**
```python
net = ps.networks.snow(im=im, voxel_size=resolution)
h = pn.check_network_health()
op.topotools.trim(network=pn, pores=h['trim_pores'])
h = pn.check_network_health()
print(h)
```
**I changed**
```python
snow = ps.networks.snow2(im, voxel_size=resolution)
pn = op.io.network_from_porespy(snow.network)

h = op.utils.check_network_health(pn)
op.topotools.trim(network=pn, pores=h['disconnected_pores'])
h = op.utils.check_network_health(pn)
print(h)
```

**Comment by JahanzebTariq** [comment]
when I pass these arguments in the Snow2 algorithm I get on x direction of permeability 

```python
snow = ps.networks.snow2(im, voxel_size=data['resolution'], boundary_width=[3, 0, 0], accuracy='standard')
```

This piece of code is taken from your recommended script.

**Reply by jgostick** [reply]
I'm not sure what you're question is.  Putting boundary_width of [3, 0, 0] means that boundary pores are only on the x faces.  If you want to do the other directions you can put boundary pores there too.  Of course you can also put no boundary pores and just use real pores on the faces for assigning boundary conditions.  Ultimately, the user is responsible for making these decisions.

**Reply by JahanzebTariq** [reply]
@jgostick .. and sir please show how can I remove the boundary pores..?

**Reply by JahanzebTariq** [reply]
There has been alot of change in the SNOW algorithm particularly naming, which is confusing me a lot, making me fearful of making mistakes..
